java_gen/templates/custom/interface/OFMessage.java

```diff
@@ -0,0 +1,15 @@
+    
+    /**
+     * Compares the two messages for equality, ignoring the XID field.
+     *
+     * @param obj the other message to compare
+     * @return true if the messages are equal, ignoring the XID; false otherwise
+     */
+    boolean equalsIgnoreXid(Object obj);
+
+    /**
+     * Computes the hashcode of the message, ignoring the XID field.
+     *
+     * @return the hashcode of the message, ignoring the XID
+     */
+    int hashCodeIgnoreXid();
```

Comment:
@andi-bigswitch Good question. In Floodlight, we have a cache for OFMessages that have been sent to particular switches. This cache is maintained to avoid sending duplicate messages within a very short period. An easy-to-understand use-case for this cache is when inserting flows reactively. If a UDP connection starts, the controller could potentially receive a large number of UDP packets as OFPacketIns, triggering OFFlowMods. The controller might receive a large number of the UDP packets prior to actually having the flow installed to handle these packets in the data plane. Without this cache, many identical OFFlowMod messages could be sent to the switch.

The cache is implemented using a map, where the key to the map is an (OFMessage, IOFSwitch) tuple -- to temporarily store which messages were sent on which switches within a certain time period. Without a custom hash function that ignores the XID, the hash of our tuple class is a combination of the OFMessage hash and the IOFSwitch hash, which will produce a different hash for identical OFMessages with different XIDs. But, the XIDs *should* be different and thus the cache won't work and a hit will never occur for two identical messages (w/different XIDs) sent to the same switch.

The solution is to get a hash for the OFMessage that omits the XID hash in its hash computation. This hash function -- hashCodeIgnoreXid() -- is explicitly called within the hash computation of our (OFMessage, IOFSwitch) tuple class for the OFMessage member instead of the default hashCode(). This is far more efficient than implementing a brute-force equalsIgnoreXid() comparison as a way to check the keys in the map.